src/node/app.ts

```diff
@@ -22,39 +22,33 @@ export interface App extends Disposable {
   server: http.Server
 }
 
-const listen = (server: http.Server, { host, port, socket, "socket-mode": mode }: ListenOptions) => {
-  return new Promise<void>(async (resolve, reject) => {
+export const listen = async (server: http.Server, { host, port, socket, "socket-mode": mode }: ListenOptions) => {
+  if (socket) {
+    try {
+      await fs.unlink(socket)
+    } catch (error: any) {
+      handleArgsSocketCatchError(error)
+    }
+  }
+  await new Promise<void>(async (resolve, reject) => {
     server.on("error", reject)
-
     const onListen = () => {
       // Promise resolved earlier so this is an unrelated error.
       server.off("error", reject)
       server.on("error", (err) => util.logError(logger, "http server error", err))
-
-      if (socket && mode) {
-        fs.chmod(socket, mode)
-          .then(resolve)
-          .catch((err) => {
-            util.logError(logger, "socket chmod", err)
-            reject(err)
-          })
-      } else {
-        resolve()
-      }
+      resolve()
     }
-
     if (socket) {
-      try {
-        await fs.unlink(socket)
-      } catch (error: any) {
-        handleArgsSocketCatchError(error)
-      }
-
       server.listen(socket, onListen)
     } else {
       // [] is the correct format when using :: but Node errors with them.
       server.listen(port, host.replace(/^\[|\]$/g, ""), onListen)
     }
+    if (socket && mode) {
+      await fs.chmod(socket, mode).catch((err) => {
+        util.logError(logger, "socket chmod", err)
+      })
+    }
   })
```

Comment:
I think this may need to go outside the promise since we need to `chmod` after the server is finished listening otherwise the socket may not have been created yet.  Something like this although I not sure I quite have it right, kinda hard to edit in GitHub :stuck_out_tongue: 

```suggestion
  })
  if (socket && mode) {
    await fs.chmod(socket, mode).catch((err) => {
      util.logError(logger, "socket chmod", err)
    })
  }
```